File: macos/Murmur/Detection/Sessionizer.swift
import Foundation
import CryptoKit

class Sessionizer {
    private var currentSession: DetectedSession?
    private var isInAISession: Bool = false
    private var lastEventTime: Date?
    private let idGenerator: () -> String
    private let timezone: TimeZone
    private let dateFormatter: ISO8601DateFormatter
    private let localDateFormatter: DateFormatter
    private let deviceId: String

    init(idGenerator: @escaping () -> String = { UUID().uuidString }) {
        self.idGenerator = idGenerator
        self.timezone = TimeZone.current
        self.dateFormatter = ISO8601DateFormatter()
        self.dateFormatter.timeZone = timezone
        self.localDateFormatter = DateFormatter()
        self.localDateFormatter.dateFormat = "yyyy-MM-dd"
        self.localDateFormatter.timeZone = timezone
        self.deviceId = Sessionizer.loadOrCreateDeviceId()
    }

    /// Load or create a per-installation device identifier stored in UserDefaults.
    static func loadOrCreateDeviceId() -> String {
        let key = "murmur_device_id"
        if let existing = UserDefaults.standard.string(forKey: key), !existing.isEmpty {
            return existing
        }
        let newId = UUID().uuidString
        UserDefaults.standard.set(newId, forKey: key)
        return newId
    }

    /// Compute a source fingerprint for idempotent dedup using SHA-256.
    private func computeFingerprint(session: DetectedSession) -> String {
        let raw = [
            session.sourcePlatform.rawValue,
            session.sourceKind.rawValue,
            deviceId,
            session.toolId ?? "",
            session.rawBundleId ?? session.rawDomain ?? "",
            String(Int(session.startedAt.timeIntervalSince1970 / 5)),
            String(Int(session.endedAt.timeIntervalSince1970 / 5)),
            String(session.activeSeconds)
        ].joined(separator: "|")

        let data = Data(raw.utf8)
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    // MARK: - Process Event

    /// Process a new RawEvent through the sessionizer.
    /// Returns an array of completed sessions (may be empty, one, or two when crossing midnight).
    func processEvent(_ event: RawEvent, matchResult: ToolMatchResult) -> [DetectedSession] {
        let now = event.timestamp
        var completedSessions: [DetectedSession] = []

        // Check idle threshold (5 minutes) — flush any active session first
        if let lastTime = lastEventTime, now.timeIntervalSince(lastTime) > 300, isInAISession {
            let flushed = flushCurrentSession(at: lastTime)
            completedSessions.append(contentsOf: flushed)
            isInAISession = false
            currentSession = nil
        }

        lastEventTime = now

        let isAIEvent = matchResult.matchedTool != nil && !matchResult.shouldIgnore && matchResult.confidence >= 0.4

        if isAIEvent && !isInAISession {
            // Start new AI session. Flush any prior session first.
            let flushed = flushCurrentSession(at: now)
            completedSessions.append(contentsOf: flushed)
            startNewSession(event: event, matchResult: matchResult, at: now)
            isInAISession = true

        } else if isAIEvent && isInAISession {
            // Check if tool changed — if so, flush current and start new session
            if let currentToolId = currentSession?.toolId,
               let newToolId = matchResult.matchedTool?.id,
               currentToolId != newToolId {
                let flushed = flushCurrentSession(at: now)
                completedSessions.append(contentsOf: flushed)
                startNewSession(event: event, matchResult: matchResult, at: now)
                isInAISession = true
                return completedSessions
            }

            // Continuing same AI session — update end time
            currentSession?.endedAt = now
            currentSession?.activeSeconds = max(0, Int(now.timeIntervalSince(currentSession!.startedAt)))

        } else if !isAIEvent && isInAISession {
            // Non-AI foreground — end session
            let flushed = flushCurrentSession(at: now)
            completedSessions.append(contentsOf: flushed)
            isInAISession = false
            currentSession = nil
        }

        return completedSessions
    }

    // MARK: - Flush Current Session

    /// End the current session and return it (or two sessions if split at midnight).
    func flushCurrentSession(at endTime: Date? = nil) -> [DetectedSession] {
        guard var session = currentSession else { return [] }

        let end = endTime ?? Date()
        session.endedAt = end
        session.activeSeconds = max(0, Int(end.timeIntervalSince(session.startedAt)))

        let startDateStr = localDateFormatter.string(from: session.startedAt)
        let endDateStr = localDateFormatter.string(from: end)

        if startDateStr != endDateStr {
            // Cross-midnight: split into two sessions
            return splitCrossMidnight(session: session, startDateStr: startDateStr, endDateStr: endDateStr, actualEnd: end)
        }

        // Single date — finalize normally
        session.localDate = startDateStr
        session.isNight = isNightHours(session.startedAt) || isNightHours(end)
        session.updatedAt = Date()

        return finalize(session)
    }

    // MARK: - Cross-Midnight Split

    private func splitCrossMidnight(session: DetectedSession, startDateStr: String, endDateStr: String, actualEnd: Date) -> [DetectedSession] {
        // Calculate midnight boundary (00:00:00 of end date in local timezone)
        var calendar = Calendar.current
        calendar.timeZone = timezone
        let endDate = localDateFormatter.date(from: endDateStr)!
        let midnight = calendar.startOfDay(for: endDate)

        let firstSegmentSeconds = max(0, Int(midnight.timeIntervalSince(session.startedAt)))
        let secondSegmentSeconds = max(0, Int(actualEnd.timeIntervalSince(midnight)))

        var results: [DetectedSession] = []

        // First segment (before midnight)
        if firstSegmentSeconds >= 15 {
            var first = session
            first.id = idGenerator() // new ID for the split segment
            first.endedAt = midnight
            first.activeSeconds = firstSegmentSeconds
            first.localDate = startDateStr
            first.isNight = isNightHours(session.startedAt) || isNightHours(midnight.addingTimeInterval(-1))
            first.updatedAt = Date()
            if firstSegmentSeconds < 30 { first.status = .suspected }
            if let finalized = finalizeSingle(first) {
                results.append(finalized)
            }
        }

        // Second segment (after midnight)
        if secondSegmentSeconds >= 15 {
            var second = session
            second.id = idGenerator()
            second.startedAt = midnight
            second.endedAt = actualEnd
            second.activeSeconds = secondSegmentSeconds
            second.localDate = endDateStr
            second.isNight = isNightHours(midnight) || isNightHours(actualEnd)
            second.updatedAt = Date()
            if secondSegmentSeconds < 30 { second.status = .suspected }
            if let finalized = finalizeSingle(second) {
                results.append(finalized)
            }
        }

        // If both segments too short, keep original as one session on start date
        if results.isEmpty {
            var single = session
            single.localDate = startDateStr
            single.isNight = isNightHours(session.startedAt) || isNightHours(actualEnd)
            single.updatedAt = Date()
            if let finalized = finalizeSingle(single) {
                results.append(finalized)
            }
        }

        currentSession = nil
        return results
    }

    // MARK: - Finalize

    /// Apply noise filtering and status assignment. Returns nil if session should be discarded.
    private func finalize(_ session: DetectedSession) -> [DetectedSession] {
        if let result = finalizeSingle(session) {
            return [result]
        }
        return []
    }

    private func finalizeSingle(_ session: DetectedSession) -> DetectedSession? {
        var s = session

        // Discard sessions < 15 seconds
        if s.activeSeconds < 15 {
            return nil
        }

        // Flag 15-30s as suspected
        if s.activeSeconds < 30 {
            s.status = .suspected
        }

        s.updatedAt = Date()
        s.sourceFingerprint = computeFingerprint(session: s)
        return s
    }

    // MARK: - Private Helpers

    private func startNewSession(event: RawEvent, matchResult: ToolMatchResult, at timestamp: Date) {
        let sessionId = idGenerator()
        let now = Date()

        var session = DetectedSession(
            id: sessionId,
            sourcePlatform: event.platform,
            sourceKind: event.platform == .browser ? .web : .app,
            detectorId: "macos.workspace",
            toolId: matchResult.matchedTool?.id,
            toolName: matchResult.matchedTool?.name,
            rawAppName: event.appName,
            rawBundleId: event.bundleId,
            rawPackageName: event.packageName,
            rawDomain: event.domain,
            rawUrlPattern: event.urlPattern,
            windowTitleHash: event.windowTitleHash,
            startedAt: timestamp,
            endedAt: timestamp,
            activeSeconds: 0,
            idleSeconds: 0,
            localDate: localDateFormatter.string(from: timestamp),
            timezone: timezone.identifier,
            isNight: isNightHours(timestamp),
            confidence: matchResult.confidence,
            status: .pending,
            mergedIntoSessionId: nil,
            promptCount: 0,
            deviceId: deviceId,
            sourceSessionId: nil,
            sourceFingerprint: nil,
            syncStatus: .localOnly,
            syncedAt: nil,
            createdAt: now,
            updatedAt: now
        )

        if matchResult.needsConfirmation {
            session.status = .suspected
        }

        currentSession = session
    }

    private func isNightHours(_ date: Date) -> Bool {
        let calendar = Calendar.current
        let components = calendar.dateComponents([.hour], from: date)
        guard let hour = components.hour else { return false }
        return hour >= 22 || hour < 6
    }

    // MARK: - Merge Adjacent Sessions

    func mergeAdjacentSessions(_ sessions: [DetectedSession]) -> [DetectedSession] {
        guard sessions.count > 1 else { return sessions }

        let sorted = sessions.sorted { $0.startedAt < $1.startedAt }
        var merged: [DetectedSession] = []
        var current = sorted[0]

        for i in 1..<sorted.count {
            let next = sorted[i]
            let sameTool = current.toolId == next.toolId
            let gap = next.startedAt.timeIntervalSince(current.endedAt)
            let withinThreshold = gap <= 180 && gap >= 0

            if sameTool && withinThreshold && current.status != .completed && next.status != .completed {
                var updated = current
                updated.endedAt = max(current.endedAt, next.endedAt)
                updated.activeSeconds = Int(updated.endedAt.timeIntervalSince(updated.startedAt))
                updated.updatedAt = Date()
                current = updated
            } else {
                merged.append(current)
                current = next
            }
        }

        merged.append(current)
        return merged
    }

    // MARK: - State

    var hasActiveSession: Bool { currentSession != nil }

    func getCurrentSession() -> DetectedSession? { currentSession }

    func reset() {
        currentSession = nil
        isInAISession = false
        lastEventTime = nil
    }
}

File: android/app/src/main/java/com/murmur/app/domain/detection/Sessionizer.kt
package com.murmur.app.domain.detection

import android.content.Context
import android.content.SharedPreferences
import com.murmur.app.domain.model.*
import java.security.MessageDigest
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.util.UUID

/**
 * Converts raw usage events into detected sessions.
 * Aligned with canonical detected-session schema — generates UUID, fingerprint, timezone, etc.
 *
 * Rules:
 * - < 15s: discard
 * - 15-30s: suspect
 * - > 30s: pending
 * - 3-min merge window
 * - Cross-midnight split
 */
class Sessionizer(context: Context) {

    companion object {
        private const val MIN_SESSION_SECONDS = 15L
        private const val SUSPECTED_THRESHOLD_SECONDS = 30L
        private const val MERGE_WINDOW_MS = 3 * 60 * 1000L
        private const val PREFS_NAME = "murmur_sessionizer"
        private const val KEY_OPEN_STATE = "open_foreground_state"
        private const val KEY_DEVICE_ID = "murmur_device_id"
    }

    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val deviceId: String by lazy { loadOrCreateDeviceId() }
    private val appContext = context.applicationContext

    /**
     * Load persisted open foreground state (survives across Worker runs).
     */
    fun loadOpenForegroundState(): OpenForegroundState? {
        val json = prefs.getString(KEY_OPEN_STATE, null) ?: return null
        return try {
            val parts = json.split("::")
            if (parts.size >= 3) {
                OpenForegroundState(
                    packageName = parts[0],
                    startedAt = parts[1].toLong(),
                    toolId = parts[2]
                )
            } else null
        } catch (_: Exception) { null }
    }

    /**
     * Save open foreground state for next Worker run.
     */
    fun saveOpenForegroundState(state: OpenForegroundState?) {
        if (state == null) {
            prefs.edit().remove(KEY_OPEN_STATE).apply()
        } else {
            val json = "${state.packageName}::${state.startedAt}::${state.toolId}"
            prefs.edit().putString(KEY_OPEN_STATE, json).apply()
        }
    }

    /**
     * Process a list of RawEvents and convert them into DetectedSessions.
     * Handles persisted open foreground state from previous worker runs.
     */
    fun processEvents(
        events: List<RawEvent>,
        matcher: ToolMatcher
    ): List<DetectedSession> {
        if (events.isEmpty() && loadOpenForegroundState() == null) return emptyList()

        // Load persisted open state and merge into events
        val openState = loadOpenForegroundState()
        val allEvents = mutableListOf<RawEvent>()
        if (openState != null) {
            // Insert the open foreground as a starting event
            allEvents.add(
                RawEvent(
                    packageName = openState.packageName,
                    eventType = RawEventType.FOREGROUND,
                    timestamp = openState.startedAt
                )
            )
        }
        allEvents.addAll(events)

        // Step 1: Pair foreground/background events
        val rawSessions = pairEvents(allEvents)

        // Step 2: Match each raw session to a tool
        val matchedSessions = rawSessions.mapNotNull { rawSession ->
            val matchResult = matcher.match(rawSession.foregroundEvent)
            if (matchResult.ignored || matchResult.tool == null) {
                null
            } else {
                val durationMs = rawSession.backgroundEvent?.timestamp?.minus(rawSession.foregroundEvent.timestamp)
                    ?: (System.currentTimeMillis() - rawSession.foregroundEvent.timestamp).coerceAtLeast(0)
                val durationSeconds = durationMs / 1000

                if (durationSeconds < MIN_SESSION_SECONDS) return@mapNotNull null

                val status = when {
                    !matchResult.needsConfirmation && durationSeconds >= SUSPECTED_THRESHOLD_SECONDS -> SessionStatus.PENDING
                    else -> SessionStatus.SUSPECTED
                }

                val startedAt = rawSession.foregroundEvent.timestamp
                val endedAt = rawSession.backgroundEvent?.timestamp ?: (startedAt + durationMs)
                val now = System.currentTimeMillis()
                val zoneId = ZoneId.systemDefault()
                val startInstant = Instant.ofEpochMilli(startedAt)
                val localDate = LocalDate.ofInstant(startInstant, zoneId).toString()
                val hour = startInstant.atZone(zoneId).toLocalTime().hour
                val isNight = hour >= 22 || hour < 6

                val canonicalId = UUID.randomUUID().toString()

                DetectedSession(
                    id = 0,
                    canonicalId = canonicalId,
                    sourcePlatform = SourcePlatform.ANDROID,
                    sourceKind = SourceKind.APP,
                    detectorId = "android.usagestats",
                    toolId = matchResult.tool!!.id,
                    toolName = matchResult.tool!!.name,
                    rawAppName = getAppLabel(rawSession.foregroundEvent.packageName),
                    packageName = rawSession.foregroundEvent.packageName,
                    rawPackageName = rawSession.foregroundEvent.packageName,
                    rawDomain = null,
                    rawUrlPattern = null,
                    detectedAt = now,
                    startedAt = startedAt,
                    endedAt = endedAt,
                    activeSeconds = durationSeconds,
                    idleSeconds = 0,
                    localDate = localDate,
                    timezone = zoneId.id,
                    isNight = isNight,
                    status = status,
                    confidence = matchResult.confidence,
                    mergedIntoSessionId = null,
                    promptCount = null,
                    sourceFingerprint = null, // set after merge/split
                    deviceId = deviceId,
                    syncStatus = "local_only",
                    createdAt = now,
                    updatedAt = now
                )
            }
        }

        // Step 3: Merge adjacent
        val mergedSessions = mergeNearbySessions(matchedSessions)

        // Step 4: Split cross-midnight
        val splitSessions = splitCrossMidnight(mergedSessions)

        // Step 5: Compute fingerprints
        val finalized = splitSessions.map { s ->
            s.copy(sourceFingerprint = computeFingerprint(s))
        }

        // Update open foreground state: if last event was foreground without background, persist
        val lastEvent = events.lastOrNull()
        if (lastEvent != null && lastEvent.eventType == RawEventType.FOREGROUND) {
            val match = matcher.match(lastEvent)
            if (match.tool != null && !match.ignored) {
                saveOpenForegroundState(
                    OpenForegroundState(
                        packageName = lastEvent.packageName,
                        startedAt = lastEvent.timestamp,
                        toolId = match.tool.id
                    )
                )
            } else {
                saveOpenForegroundState(null)
            }
        } else {
            saveOpenForegroundState(null)
        }

        return finalized
    }

    /**
     * Pair FOREGROUND and BACKGROUND events into raw sessions.
     * Incorporates persisted open state.
     */
    private fun pairEvents(events: List<RawEvent>): List<RawSession> {
        val sessions = mutableListOf<RawSession>()
        val sorted = events.sortedBy { it.timestamp }

        var pendingForeground: RawEvent? = null

        for (event in sorted) {
            when (event.eventType) {
                RawEventType.FOREGROUND -> {
                    if (pendingForeground != null) {
                        // Close previous foreground at new foreground's time to prevent over-counting
                        val closureEvent = RawEvent(
                            packageName = pendingForeground.packageName,
                            eventType = RawEventType.BACKGROUND,
                            timestamp = event.timestamp
                        )
                        sessions.add(RawSession(pendingForeground, closureEvent))
                    }
                    pendingForeground = event
                }
                RawEventType.BACKGROUND -> {
                    if (pendingForeground != null &&
                        pendingForeground.packageName == event.packageName) {
                        sessions.add(RawSession(pendingForeground, event))
                        pendingForeground = null
                    }
                }
            }
        }

        // Persist remaining unclosed foreground as open state for next run
        // (it will be paired when the next worker run sees the background or next foreground)

        return sessions
    }

    private fun mergeNearbySessions(sessions: List<DetectedSession>): List<DetectedSession> {
        if (sessions.isEmpty()) return emptyList()
        if (sessions.size == 1) return sessions

        val sorted = sessions.sortedBy { it.startedAt }
        val merged = mutableListOf<DetectedSession>()
        var current = sorted.first()

        for (i in 1 until sorted.size) {
            val next = sorted[i]
            val isSameTool = current.toolId == next.toolId
            val isWithinWindow = next.startedAt - current.endedAt <= MERGE_WINDOW_MS

            if (isSameTool && isWithinWindow) {
                // Sum active seconds only — gap time is NOT active usage
                val mergedActiveSeconds = current.activeSeconds + next.activeSeconds
                current = current.copy(
                    endedAt = next.endedAt,
                    activeSeconds = mergedActiveSeconds,
                    confidence = maxOf(current.confidence, next.confidence),
                    updatedAt = System.currentTimeMillis()
                )
            } else {
                merged.add(current)
                current = next
            }
        }
        merged.add(current)
        return merged
    }

    private fun splitCrossMidnight(sessions: List<DetectedSession>): List<DetectedSession> {
        val result = mutableListOf<DetectedSession>()

        for (session in sessions) {
            val startInstant = Instant.ofEpochMilli(session.startedAt)
            val endInstant = Instant.ofEpochMilli(session.endedAt)
            val zoneId = ZoneId.of(session.timezone)
            val startDate = LocalDate.ofInstant(startInstant, zoneId)
            val endDate = LocalDate.ofInstant(endInstant, zoneId)

            if (startDate == endDate) {
                result.add(session)
            } else {
                val midnightMillis = endDate.atStartOfDay(zoneId).toInstant().toEpochMilli()

                val beforeMidnightSeconds = (midnightMillis - session.startedAt) / 1000
                val afterMidnightSeconds = (session.endedAt - midnightMillis) / 1000

                val now = System.currentTimeMillis()

                if (beforeMidnightSeconds >= MIN_SESSION_SECONDS) {
                    val newId = UUID.randomUUID().toString()
                    result.add(
                        session.copy(
                            id = 0,
                            canonicalId = newId,
                            endedAt = midnightMillis,
                            activeSeconds = beforeMidnightSeconds,
                            localDate = startDate.toString(),
                            isNight = isNightHour(zoneId, midnightMillis - 1000),
                            status = if (beforeMidnightSeconds < SUSPECTED_THRESHOLD_SECONDS)
                                SessionStatus.SUSPECTED else session.status,
                            updatedAt = now,
                            sourceFingerprint = null // recomputed later
                        )
                    )
                }

                if (afterMidnightSeconds >= MIN_SESSION_SECONDS) {
                    val newId = UUID.randomUUID().toString()
                    result.add(
                        session.copy(
                            id = 0,
                            canonicalId = newId,
                            startedAt = midnightMillis,
                            activeSeconds = afterMidnightSeconds,
                            localDate = endDate.toString(),
                            isNight = isNightHour(zoneId, midnightMillis),
                            status = if (afterMidnightSeconds < SUSPECTED_THRESHOLD_SECONDS)
                                SessionStatus.SUSPECTED else session.status,
                            updatedAt = now,
                            sourceFingerprint = null
                        )
                    )
                }
            }
        }

        return result
    }

    private fun isNightHour(zoneId: ZoneId, epochMs: Long): Boolean {
        val hour = Instant.ofEpochMilli(epochMs).atZone(zoneId).toLocalTime().hour
        return hour >= 22 || hour < 6
    }

    private fun computeFingerprint(session: DetectedSession): String {
        val parts = listOf(
            session.sourcePlatform.value,
            session.sourceKind.value,
            deviceId,
            session.toolId,
            session.rawPackageName ?: session.packageName,
            (session.startedAt / 5000).toString(),
            (session.endedAt / 5000).toString(),
            session.activeSeconds.toString()
        ).joinToString("|")

        val digest = MessageDigest.getInstance("SHA-256").digest(parts.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }

    private fun loadOrCreateDeviceId(): String {
        val existing = prefs.getString(KEY_DEVICE_ID, null)
        if (!existing.isNullOrEmpty()) return existing
        val newId = UUID.randomUUID().toString()
        prefs.edit().putString(KEY_DEVICE_ID, newId).apply()
        return newId
    }

    private fun getAppLabel(packageName: String): String? {
        return try {
            val pm = appContext.packageManager
            val appInfo = pm.getApplicationInfo(packageName, 0)
            pm.getApplicationLabel(appInfo).toString()
        } catch (_: Exception) { null }
    }

    data class OpenForegroundState(
        val packageName: String,
        val startedAt: Long,
        val toolId: String
    )

    private data class RawSession(
        val foregroundEvent: RawEvent,
        val backgroundEvent: RawEvent?
    )
}
